File: fireballpy/fireball.py
from __future__ import annotations
from copy import deepcopy
import warnings

from numpy.typing import ArrayLike
import numpy as np

from fireballpy._errors import raise_fb_error, type_check
from fireballpy._options import get_icharge, get_idipole, get_imixer
from fireballpy._correction import Correction
from fireballpy.fdata import FDataFiles
from fireballpy.atoms import AtomSystem
from fireballpy.kpoints import KPoints

from ase import Atoms

from fireballpy._fireball import (set_options,
                                  set_initial_charges,
                                  call_allocate_system,
                                  scf,
                                  get_sizes,
                                  get_initial_charges,
                                  calc_forces)

DEFAULT_MIXER = {'method': 'johnson',
                 'max_iter': np.int64(200),
                 'mix_order': np.int64(6),
                 'beta': np.float64(0.1),
                 'tol': np.float64(1e-8),
                 'w0': np.float64(0.01)}


class BaseFireball:
    """Base Fireball class to be interfaced with any high-level package such as ASE.

    Parameters
    ----------
    atomsystem : AtomSystem
        Class with all the information of species, atomic numbers,
        positions and cell of the system.
    fdatafiles : FDataFiles
        Class with all the information on the location of the basis functions.
    kpoints : KPoints
        Class with all the information of the coordinates of the k-points as well
        as their associated weights.
    verbose : bool, optional
        If ``True`` information of the convergence of the SCF loop will be printed on screen.
        Default is ``False``.
    total_charge : int, optional
        Total charge of the system in elementary charge units (1 extra electron would be ``total_charge = -1``).
        By default this is 0.
    correction : dict | bool | None, optional
        By default (``None``) will apply DFT3 correction for the selected
        FData if optimized parameters are available
        (see ``available_fdatas()``).
        Also, it may be a dictionary with the parameters for DFTD3 correction
        The dictionary must contain the keys ``'kind'`` (right now it can only
        be ``'dftd3'``), ``'damping'`` and either ``'method'`` or
        ``'params_tweaks'``. It may be forcefully shut down by providing ``False``.
        For more information, see
        `Simple DFT-D3 documentation <https://dftd3.readthedocs.io>`_.
    charges_method : str | None, optional
        How the autoconsistency in the charges will be performed.
        By default depends on the FData (``None``).
        If a custom FData is selected, then this parameter must be specified.
        For more information see :ref:`here <charges_methods>`.
    dipole_method : str, optional
        Whether to use (``dipole='improved'``) or not (``dipole='legacy'``)
        the improved dipole description.
        By default ``'improved'`` except for periodic systems as it is not yet implemented and thus it will be ignored.
    initial_charges : ArrayLike[float], optional
        A (natoms, nshells) array with the charges of each of the atom shells (0 if no shell for that atom is defined).
    mixer_kws : dict, optional
        Dictionary with the mixer parameters.
        More information :ref:`here <mixer>`.

    Methods
    -------
    run_scf(fix_charges=False)
        Run the SCF loop and store energy and charge related quantities.
    calc_forces(fix_charges=False)
        Compute the forces and store them.
    update_coords(positions)
        Update the coordinates in the Fortran module.

    Notes
    -----
    If you use this code in a publication please cite us [Lewis2011]_.

    .. [Lewis2011] James. P. Lewis, et al.,
                   "Advances and applications in the FIREBALL ab initio
                   tight-binding molecular-dynamics formalism",
                   *Phys. Status Solidi B*, 248(9):1989-2007, 2011.
                   `DOI:10.1002/pssb.201147259
                   <https://doi.org/10.1002/pssb.201147259>`_

    """
    def __init__(self, *,
                 atomsystem: AtomSystem,
                 fdatafiles: FDataFiles,
                 kpoints: KPoints,
                 verbose: bool = False,
                 total_charge: int = 0,
                 correction: dict | None = None,
                 charges_method: str | None = None,
                 dipole_method: str = 'improved',
                 initial_charges: ArrayLike | None = None,
                 mixer_kws: dict | None = None) -> None:

        # Init variables
        self.scf_computed = False
        self.forces_computed = False

        # Type safety
        type_check(atomsystem, AtomSystem, 'atomsystem')
        type_check(fdatafiles, FDataFiles, 'fdatafiles')
        type_check(kpoints, KPoints, 'kpoints')
        type_check(verbose, bool, 'verbose')
        type_check(total_charge, int, 'total_charge')
        type_check(dipole_method, str, 'dipole_method')


        # Save everything to self
        self.fdatafiles = fdatafiles
        self.atomsystem = atomsystem
        self.kpoints = kpoints
        self.verbose = verbose
        self.total_charge = total_charge
        self.charges_method = charges_method
        self.dipole_method = dipole_method

        # Convenient aliases
        self.natoms = self.atomsystem.n
        self.nkpts = self.kpoints.n

        # Get charges method
        if self.charges_method is None:
            self.charges_method = self.fdatafiles.get_charges_method()
        else:
            type_check(self.charges_method, str, 'charges_method')
            self.charges_method = self.charges_method.lower()

        # Address correction
        self.correction = None
        if correction:
            type_check(correction, dict, 'correction', ' if custom parameters are desired')
            self._correction = Correction(atomsystem=self.atomsystem, **correction)
            self.correction = correction
        elif correction is None:
            self.correction = self.fdatafiles.get_correction(self.charges_method)
            if self.correction:
                self._correction = Correction(atomsystem=self.atomsystem, **self.correction)

        # Address mixer
        self.mixer_kws = deepcopy(DEFAULT_MIXER)
        if mixer_kws is not None:
            type_check(mixer_kws, dict, 'mixer_kws')
            try:
                self.mixer_kws = {'method': mixer_kws['method']}
            except KeyError:
                raise ValueError("Parameter ``mixer_kws`` if specified must contain the key ``'method'``.")
            for prop in ['max_iter', 'mix_order', 'beta', 'tol', 'w0']:
                self.mixer_kws[prop] = mixer_kws.get(prop, DEFAULT_MIXER[prop])

        # Save variables to modules
        self.fdatafiles.load_fdata()
        self.atomsystem.set_coords()
        self.kpoints.set_kpoints()
        self.atomsystem.set_cell()

        # Set Fireball-like options
        self._options = {'dipole_method': np.int64(0) if self.atomsystem.isperiodic else get_idipole(dipole_method),
                         'charges_method': get_icharge(self.charges_method),
                         'ismolecule': np.int64(not self.atomsystem.isperiodic),
                         'isgamma': np.int64(self.kpoints.isgamma),
                         'total_charge': np.int64(-self.total_charge),
                         'mixer_method': get_imixer(self.mixer_kws['method']),
                         'max_iter': np.int64(self.mixer_kws['max_iter']),
                         'mix_order': np.int64(self.mixer_kws['mix_order']),
                         'beta': np.float64(self.mixer_kws['beta']),
                         'tol': np.float64(self.mixer_kws['tol']),
                         'w0': np.float64(self.mixer_kws['w0'])}
        set_options(**self._options)

        # Allocate module
        call_allocate_system()
        self.nbands = 0
        self._alloc_arrays()
        if initial_charges is not None:
            if not isinstance(initial_charges, (tuple, list, np.ndarray)):
                raise ValueError("Parameter ``initial_charges`` need to be coerced into an array.")
            self.initial_charges = np.ascontiguousarray(initial_charges, dtype=np.float64)
            if self.initial_charges.shape != (self.natoms, self.nshells):
                raise ValueError("Parameter ``initial_charges`` must be a natoms x nshells array. "
                                 f"Got shape {self.initial_charges.shape}")
            set_initial_charges(self.initial_charges.T)
        else:
            self.initial_charges = np.zeros((self.natoms, self.nshells), dtype=np.float64, order='C')
            get_initial_charges(self.initial_charges.T)

    def _alloc_arrays(self) -> None:
        self.nshells, self.nbands_new, self.norbitals = get_sizes()
        if self.nbands_new != self.nbands:
            self.nbands = self.nbands_new
            self.charges = np.zeros(self.natoms, dtype=np.float64, order='C')
            self.forces = np.zeros((self.natoms, 3), dtype=np.float64, order='C')
            self.shell_charges = np.zeros((self.natoms, self.nshells), dtype=np.float64, order='C')
            self.eigenvalues = np.zeros((self.nkpts, self.nbands), dtype=np.float64, order='C')
            self.eigenvectors = np.zeros((self.nkpts, self.norbitals, self.norbitals), dtype=np.complex128, order='C')

    def run_scf(self, fix_charges: bool = False) -> None:
        """Check if SCF is computed, execute the loop if not,
        then store energy, fermi_level, partial charges, shell charges and eigenvalues.

        Parameters
        ----------
        fix_charges : bool
            If ``True`` then only one SCF loop iteration will be performed. If ``False`` (default)
            then it will perform the usual computation.
            In general it is useful to fix the charges when making postprocessing.
        """
        if not self.scf_computed:
            converged, fb_errno, energy, fermi_level, charges = scf(self.verbose,
                                                                    fix_charges,
                                                                    self.shell_charges.T,
                                                                    self.eigenvalues.T,
                                                                    self.eigenvectors.T)
            if fb_errno != 0:
                raise_fb_error(fb_errno)
            if not converged:
                warnings.warn("SCF loop did not converge. Try incrementing ``'max_iter'`` in ``mixer_kws``", UserWarning)
            self.energy = float(energy)
            self.fermi_level = float(fermi_level)
            self.charges = charges
            self.scf_computed = True
            if self.correction:
                self._correction.correct()  # type: ignore

    def calc_forces(self, fix_charges: bool = False) -> None:
        """Check if SCF is computed, execute the loop if not,
        then compute and store forces.

        Parameters
        ----------
        fix_charges : bool
            If ``True`` then only one SCF loop iteration will be performed. If ``False`` (default)
            then it will perform the usual computation.
            In general it is useful to fix the charges when making postprocessing.
        """
        self.run_scf(fix_charges)
        if not self.forces_computed:
            fb_errno = calc_forces(self.forces.T)
            if fb_errno != 0:
                raise_fb_error(fb_errno)
            self.forces_computed = True

    def update_coords(self, positions: ArrayLike) -> None:
        """Update the coordinates in the module.

        Parameters
        ----------
        positions : ArrayLike[float]
            A natoms x 3 array with the new positions in angstroms.
        """
        self.atomsystem.update_coords(positions)
        if self.correction:
            self._correction.update_coords(self.atomsystem)  # type: ignore
        self.scf_computed = False
        self.forces_computed = False


def fbobj_from_obj(fbobj: BaseFireball | None, atoms: Atoms | None) -> BaseFireball:
    fbobj_none = fbobj is not None
    atoms_none = atoms is not None

    nonecount = int(fbobj_none) + int(atoms_none)
    if nonecount != 1:
        raise ValueError("Only one of the Fireball-providing inputs must be specified.")

    if fbobj_none:
        type_check(fbobj, BaseFireball, 'fbobj')
        return fbobj
    if atoms_none:
        type_check(atoms, Atoms, 'atoms')
        if not hasattr(atoms, 'calc') or not isinstance(atoms.calc, BaseFireball):
            raise ValueError("ASE calculator for parameter ``atoms`` must be ``Fireball``.")
        return atoms.calc
    raise RuntimeError("Fireball-providing inputs could not be identified.")

File: fireballpy/fireball.f90
! Set int options
subroutine set_options(dipole_method, charges_method, &
    & ismolecule, isgamma, total_charge, mixer_method, &
    & max_iter, mix_order, beta, w0, tol)
  use iso_c_binding
  use M_system, only: igamma, icluster, idipole, iqout, iqmmm, &
    & qstate, ialgmix, max_scf_iterations, idmix, w02, bmix, sigmatol
  implicit none
  integer, intent(in) :: dipole_method, charges_method, &
    & ismolecule, isgamma, total_charge, mixer_method, max_iter, mix_order
  real(c_double), intent(in) :: beta, w0, tol
  iqmmm = 0 ! ensure this is off by default
  idipole = dipole_method
  iqout = charges_method
  icluster = ismolecule
  igamma = isgamma
  qstate = total_charge
  ialgmix = mixer_method
  max_scf_iterations = max_iter
  idmix = mix_order
  bmix = beta
  w02 = w0*w0
  sigmatol = tol
end subroutine set_options

! Set cell vectors
subroutine set_cell(a1, a2, a3)
  use iso_c_binding
  use M_system, only: a1vec, a2vec, a3vec
  implicit none
  real(c_double), dimension(3), intent(in) :: a1, a2, a3
  a1vec = a1
  a2vec = a2
  a3vec = a3
end subroutine set_cell

! Set coordinates
subroutine set_coords(naux, z, xyz)
  use iso_c_binding
  use M_system, only: natoms, ratom, symbol, imass
  use M_fdata, only : symbolA, nspecies, nzx
  implicit none
  integer, intent(in) :: naux
  integer, dimension(naux), intent(in) :: z
  real(c_double), dimension(3, naux), intent(in) :: xyz
  integer :: iatom, ispec
  natoms = naux
  if (allocated(ratom)) deallocate(ratom)
  if (allocated(symbol)) deallocate(symbol)
  if (allocated(imass)) deallocate(imass)
  allocate (ratom (3, naux))
  allocate (symbol (naux))
  allocate (imass (naux))
  do iatom = 1, naux
    ratom(:, iatom) = xyz(:, iatom)
    do ispec = 1, nspecies
      if (z(iatom) .eq. nzx(ispec)) then
        imass(iatom) = ispec
        symbol(iatom) = symbolA(ispec)
        exit
      end if
    end do
  end do
end subroutine set_coords

! Provide input charges
subroutine set_initial_charges(natoms, nsh_max, qinput)
  use iso_c_binding
  use M_system, only : Qin
  implicit none
  integer, intent(in) :: nsh_max, natoms
  real(c_double), dimension(nsh_max, natoms), intent(in) :: qinput
  Qin = qinput
end subroutine set_initial_charges

subroutine set_fix_shell_charge(nsh_max, fix_shell_charge_aux)
  use iso_c_binding
  use M_system, only : fix_shell_charge
  implicit none
  integer, intent(in) :: nsh_max
  real(c_double), dimension(nsh_max), intent(in) :: fix_shell_charge_aux
  integer :: iatom, issh
  if (allocated(fix_shell_charge)) deallocate(fix_shell_charge)
  allocate (fix_shell_charge(nsh_max))
  do issh=1,nsh_max
    fix_shell_charge(issh)=fix_shell_charge_aux(issh)
  end do  
end subroutine set_fix_shell_charge


! Faster set coordinates for simulations
subroutine update_coords(natoms, xyz)
  use iso_c_binding
  use M_system, only: ratom
  implicit none
  integer, intent(in) :: natoms
  real(c_double), dimension(3, natoms), intent(in) :: xyz
  ratom = xyz
end subroutine update_coords
 
! Set kpoints as needed by Fireball
subroutine set_kpoints(nkpts, kpts, weights)
  use iso_c_binding
  use M_system, only: nkpoints, special_k, weight_k
  implicit none
  integer, intent(in) :: nkpts
  real(c_double), dimension(3, nkpts), intent(in) :: kpts
  real(c_double), dimension(nkpts), intent(in) :: weights
  if (allocated(special_k)) deallocate(special_k)
  if (allocated(weight_k)) deallocate(weight_k)
  allocate (special_k(3, nkpts))
  allocate (weight_k(nkpts))
  nkpoints = nkpts
  weight_k = weights
  special_k = kpts
end subroutine set_kpoints

! Set MM positions and charges
! positions are in Angstroms and charges in |e| units
subroutine set_qmmm(n, pos, qs, rc1, rc2, width)
  use iso_c_binding
  use M_system, only: iqmmm, qmmm_qm_natoms, qmmm_qm_xcrd, qmmm_dxyzcl, qmmm_rc1, qmmm_rc2, qmmm_width
  implicit none
  integer, intent(in) :: n
  real(c_double), dimension(3, n), intent(in) :: pos
  real(c_double), dimension(n), intent(in) :: qs
  real(c_double), intent(in) :: rc1, rc2, width
  if (allocated(qmmm_qm_xcrd)) deallocate(qmmm_qm_xcrd)
  allocate (qmmm_qm_xcrd(4, n))
  if (allocated(qmmm_dxyzcl)) deallocate(qmmm_dxyzcl)
  allocate (qmmm_dxyzcl(3, n))
  iqmmm = 1
  qmmm_qm_natoms = n
  qmmm_qm_xcrd(1:3, :) = pos
  qmmm_qm_xcrd(4, :) = qs
  qmmm_rc1 = rc1
  qmmm_rc2 = rc2
  qmmm_width = width
end subroutine set_qmmm

! Update MM positions
subroutine update_qmmm(n, pos)
  use iso_c_binding
  use M_system, only: iqmmm, qmmm_qm_xcrd
  implicit none
  integer, intent(in) :: n
  real(c_double), dimension(3, n), intent(in) :: pos
  iqmmm = 1
  qmmm_qm_xcrd(1:3, :) = pos
end subroutine update_qmmm

! Load FData in the module
subroutine loadfdata_from_path(fdata_path)
  use iso_c_binding
  use M_fdata, only: fdataLocation
  implicit none
  character(len=400), intent(in) :: fdata_path
  fdatalocation = trim(fdata_path)
  call load_fdata()
end subroutine loadfdata_from_path
 
! Allocate all needed arrays. Python gc handles deallocate after work
subroutine call_allocate_system()
  use iso_c_binding
  implicit none
  call allocate_system()
end subroutine call_allocate_system

! Return initial charges
subroutine get_initial_charges(natoms, nsh_max, qinitial)
  use iso_c_binding
  use M_fdata, only : Qneutral, nssh
  use M_system, only : imass
  implicit none
  integer, intent(in) :: nsh_max, natoms
  real(c_double), dimension(nsh_max, natoms), intent(inout) :: qinitial
  integer :: iatom, issh, in1
  do iatom=1,natoms
    in1 = imass(iatom)
    do issh=1,nssh(in1)
      qinitial(issh, iatom) = Qneutral(issh, in1)
    end do
    do issh=nssh(in1)+1,nsh_max
      qinitial(issh, iatom) = 0.0d0
    end do
  end do
end subroutine get_initial_charges

! Compute the SCF loop
subroutine scf(natoms, nshell, nkpts, norbitals, &
    & verbose, fix_charges, shell_charges, eigenvalues, eigenvectors, &
    & nbands, converged, errno_out, energy, fermi_level, charges)
  use iso_c_binding
  use M_system, only: errno, scf_achieved, etot, efermi, eigen_k, Qin, imass, ifixcharge, icluster, &
    & igamma, bbnkre, bbnkim, norbitals_new
  use M_fdata, only: Qneutral, nssh
  implicit none
  integer, intent(in) :: natoms, nshell, nkpts, norbitals
  logical, intent(in) :: verbose, fix_charges
  real(c_double), dimension(nshell, natoms), intent(inout) :: shell_charges
  real(c_double), dimension(norbitals, nkpts), intent(inout) :: eigenvalues
  complex(c_double_complex), dimension(norbitals, norbitals, nkpts), intent(inout) :: eigenvectors
  logical, intent(out) :: converged
  integer, intent(out) :: errno_out, nbands
  real(c_double), intent(out) :: energy, fermi_level
  real(c_double), dimension(natoms), intent(out) :: charges
  integer :: iatom, issh, in1
  errno = 0
  if(fix_charges) then
    ifixcharge = 1
  else
    ifixcharge = 0
  end if
  call scf_loop (verbose)
  converged = scf_achieved
  errno_out = errno
  energy = etot
  fermi_level = efermi
  do iatom = 1, natoms
    in1 = imass(iatom)
    charges(iatom) = 0.0d0
    do issh = 1, nssh(in1)
      charges(iatom) = charges(iatom) + Qneutral(issh,in1) - Qin(issh,iatom)
    end do
  end do
  shell_charges = Qin
  eigenvalues = eigen_k(:, :)
  if (icluster .eq. 0 .and. igamma .eq. 0) then
    eigenvectors = cmplx(bbnkre, bbnkim, kind=c_double_complex)
  else
    eigenvectors = cmplx(bbnkre, kind=c_double_complex)
  end if
  nbands = norbitals_new
end subroutine scf

! Get the forces in each atom
subroutine calc_forces(natoms, forces, errno_out)
  use iso_c_binding
  use M_system, only : ftot, errno
  implicit none
  integer, intent(in) :: natoms
  real(c_double), dimension(3, natoms), intent(inout) :: forces
  integer, intent(out) :: errno_out
  errno = 0
  call getforces()
  errno_out = errno
  forces = ftot
end subroutine calc_forces

! Get the forces of the QM region on the MM region
subroutine get_qmmm_forces(natoms, forces)
  use iso_c_binding
  use M_system, only : qmmm_dxyzcl
  implicit none
  integer, intent(in) :: natoms
  real(c_double), dimension(3, natoms), intent(inout) :: forces
  forces = qmmm_dxyzcl
end subroutine get_qmmm_forces

! Get hamiltonian and overlap matrix
subroutine get_orbitals(natoms, orbitals)
  use iso_c_binding
  use M_system, only : imass
  use M_fdata, only : num_orb
  implicit none
  integer, intent(in) :: natoms
  integer, dimension(natoms), intent(inout) :: orbitals
  integer :: iatom
  do iatom = 1, natoms
    orbitals(iatom) = num_orb(imass(iatom))
  end do
end subroutine get_orbitals

! Get hamiltonian and overlap matrix
subroutine get_hs(norbitals, sdat, hdat)
  use iso_c_binding
  implicit none
  integer, intent(in) :: norbitals
  real(c_double), dimension(norbitals, norbitals), intent(inout) :: sdat, hdat
  call geth(sdat, hdat)
end subroutine get_hs


